gxtest/copyfilter.cpp

```diff
@@ -0,0 +1,160 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+
+#include <ogcsys.h>
+#include <wiiuse/wpad.h>
+#include "common/hwtests.h"
+#include "gxtest/cgx.h"
+#include "gxtest/cgx_defaults.h"
+#include "gxtest/util.h"
+
+// Use all copy filter values (0-63*3), instead of just 0-64
+#define FULL_COPY_FILTER_COEFS true
+// Use all gamma values, instead of just 1.0 (0)
+#define FULL_GAMMA true
+
+static void FillEFB(u8 a, u8 r, u8 g, u8 b)
+{
+  PE_CONTROL ctrl;
+  ctrl.hex = BPMEM_ZCOMPARE << 24;
+  ctrl.pixel_format = PIXELFMT_RGB8_Z24;
+  ctrl.zformat = ZC_LINEAR;
+  ctrl.early_ztest = 0;
+  CGX_LOAD_BP_REG(ctrl.hex);
+
+  CGX_LOAD_BP_REG(BPMEM_CLEAR_AR << 24 | u32(a) << 8 | u32(r));
+  CGX_LOAD_BP_REG(BPMEM_CLEAR_GB << 24 | u32(g) << 8 | u32(b));
+
+  GXTest::CopyToTestBuffer(0, 0, 199, 49, true);
+}
+
+enum class Gamma : u8
+{
+  Gamma1_0 = 0,
+  Gamma1_7 = 1,
+  Gamma2_2 = 2,
+  Invalid2_2 = 3,  // Behaves the same as Gamma2_2?
+};
+
+#if FULL_GAMMA
+static const std::array<Gamma, 4> GAMMA_VALUES = { Gamma::Gamma1_0, Gamma::Gamma1_7, Gamma::Gamma2_2, Gamma::Invalid2_2 };
+#else
+// For now Dolphin doesn't implement gamma for EFB copies
+static const std::array<Gamma, 1> GAMMA_VALUES = { Gamma::Gamma1_0 };
+#endif
+
+#if FULL_COPY_FILTER_COEFS
+#define MAX_COPY_FILTER 63*3
+#else
+#define MAX_COPY_FILTER 64
+#endif
+void SetCopyFilter(u8 copy_filter_sum)
```

Comment:
Might be worth triggering an EFB copy from one of the YUV formats as the source format to increase the testing space, if that is supported by hardware at all.

Reply:
Yeah, the YUV/intensity formats, the depth formats, and an EFB format with alpha would all be good to test.  I am slightly concerned that the test currently takes ~10 minutes to run, and adding in more formats will only increase that time, but I can put those behind additional defines like with `FULL_COPY_FILTER_COEFS` and `FULL_GAMMA`.

Reply:
To speed up test runtime: Instead of triggering one EFB copy for each tested value, couldn't you put a different test value into each EFB pixel and then read out each output pixel? That should easily bring down the runtime to a few seconds, at which point running the whole thing again for the other formats wouldn't be such a big issue.

Reply:
That's a good idea.  I'll probably use `GX_PokeARGB` for setting it up that way, since it seems like the easiest technique.

Reply:
You could also use the `Quad` utility to draw 1x1 pixel rectangles, which should have better performance on Dolphin than EFB pokes.

Reply:
EFB pokes seem to have OK enough performance, at least in single core.  I've done some additional work, and the test currently passes on real hardware with `PIXELFMT_RGB8_Z24`, `PIXELFMT_RGBA6_Z24`, `PIXELFMT_Y8`, and `PIXELFMT_YUV420` when using EFB pokes, but fails with the other formats.  (With Dolphin, it only passes with `PIXELFMT_RGB8_Z24` currently).  What's odd is that I didn't need to do any special handling with `PIXELFMT_Y8` and `PIXELFMT_YUV420`; maybe EFB pokes are poking YUV data in that case and not RGB data?

Reply:
Ah, I forgot EFB pokes take less of a dramatic performance penalty in Dolphin these days :)

> What's odd is that I didn't need to do any special handling with PIXELFMT_Y8 and PIXELFMT_YUV420; maybe EFB pokes are poking YUV data in that case and not RGB data?

I can't remember whether EFB pokes take a pixel format into account on real hw or just write untyped data. Naively I would expect the same amount of color conversion to take place for YUV as when writing RGBA6 data. If the input data is always converted from RGBA8, then chances are the hardware path for EFB pokes doesn't implement YUV conversion for simplicity.